cli/loader/src/lib.rs

```diff
@@ -1280,6 +1247,36 @@ impl Loader {
         pattern.and_then(|r| RegexBuilder::new(r).multi_line(true).build().ok())
     }
 
+    fn grammar_json_name(grammar_path: &Path) -> Result<String> {
+        let file = fs::File::open(grammar_path).with_context(|| {
+            format!("Failed to open grammar.json at {}", grammar_path.display())
+        })?;
+
+        let first_three_lines = BufReader::new(file)
+            .lines()
+            .take(3)
+            .collect::<Result<Vec<_>, _>>()
+            .with_context(|| {
+                format!(
+                    "Failed to read the first three lines of grammar.json at {}",
+                    grammar_path.display()
+                )
+            })?
+            .join("\n");
+
+        let name = GRAMMAR_NAME_REGEX
+            .captures(&first_three_lines)
```

Comment:
I don't think that optimization is safe; the first three lines are not guaranteed to contain that name, e.g.,
https://github.com/tree-sitter-grammars/tree-sitter-slang/blob/dd991eb3b6957a33d9044e0f5914588f7f449a78/src/grammar.json#L2-L6

(Current CLI versions no longer generate such a -- bugged -- `grammar.json`, but these still exist in the wild as my crater test showed.)